### PR TITLE
Add scripts to generate plots of LRSchedulers

### LINKED ISSUE #92007: Add plots of LRSchedulers to doc to make it easier to read
We can do something similar to what we have for activation functions https://pytorch.org/docs/stable/generated/torch.nn.ReLU.html?highlight=relu#torch.nn.ReLU

A simple script similar to the one shared in https://github.com/pytorch/pytorch/issues/91765 can be used a the basis for generating these plots.

cc @svekars @carljparker

### PR DESCRIPTION
**Pull request by zeshengzong**
Fixes #92007

## Changes

- Add script to generate plots for `lr_scheduler`
- Add plots to `lr_scheduler` docs
- Add example section if it missing in `lr_scheduler` docs

### LambdaLR

image

### MultiplicativeLR

image

### StepLR

image

### MultiStepLR

image

### ConstantLR

image

### LinearLR

image

### ExponentialLR

image

### PolynomialLR

image

### CosineAnnealingLR

image

### ChainedScheduler

image

### SequentialLR

image

### ReduceLROnPlateau

image

### CyclicLR

image

### OneCycleLR

image

### CosineAnnealingWarmRestarts

image

### CODE REVIEW

**Comment by zeshengzong** [comment]
@albanD @janeyx99 Hello, please help review the change, thanks!

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -131,6 +131,9 @@ def wrapper(*args, **kwargs):
         patch_track_step_called(self.optimizer)
         self._initial_step()
 
+    def _get_name(self) -> str:
+        return self.__class__.__name__
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:135]: feel free to just call self.__class__.__name__ instead of making a new API

**Review by janeyx99 — COMMENTED** [review]
Thanks for taking this on! Very cool.

Two things:
1. Can we make sure every example in the doc aligns with the plot?
2. Can we make sure every plot shows what one would expect? e.g., CyclicLR doesn't look...cyclic haha, and PolynomialLR might benefit from a power that isn't 1, like 0.9 or something.

File: docs/source/scripts/build_lr_scheduler_images.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+model = torch.nn.Linear(10, 1)
+optimizer = optim.SGD(model.parameters(), lr=0.1)
+
+num_epochs = 100
+
+scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=2)
+scheduler2 = ExponentialLR(optimizer, gamma=0.9)
+
+schedulers = [
+    (lambda opt: LambdaLR(opt, lr_lambda=lambda epoch: 1 - epoch / num_epochs)),
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply docs/source/scripts/build_lr_scheduler_images.py]: Just a random pick, change back to previous example in doc :D

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py]: Changed, thanks!

File: docs/source/scripts/build_lr_scheduler_images.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    (lambda opt: MultiplicativeLR(opt, lr_lambda=lambda epoch: 0.95)),
+    (lambda opt: StepLR(opt, step_size=30, gamma=0.1)),
+    (lambda opt: MultiStepLR(opt, milestones=[30, 80], gamma=0.1)),
+    (lambda opt: ConstantLR(opt, factor=0.5, total_iters=4)),
+    (lambda opt: LinearLR(opt, start_factor=0.5, total_iters=4)),
+    (lambda opt: ExponentialLR(opt, gamma=0.95)),
+    (lambda opt: PolynomialLR(opt, total_iters=4, power=1.0)),
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply docs/source/scripts/build_lr_scheduler_images.py]: Make some changes to consistent doc and plot, now it look like this:

### LambdaLR

image

### MultiplicativeLR

image

### StepLR

image

### MultiStepLR

image

### ConstantLR

image

### LinearLR

image

### ExponentialLR

image

### PolynomialLR

image

### CosineAnnealingLR

image

### ChainedScheduler

image

### SequentialLR

image

### ReduceLROnPlateau

image

### CyclicLR

image

### OneCycleLR

image

### CosineAnnealingWarmRestarts

image

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -647,13 +638,7 @@ class LinearLR(LRScheduler):
 
     Example:
         >>> # xdoctest: +SKIP
-        >>> # Assuming optimizer uses lr = 0.05 for all groups
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:629]: same here

File: torch/optim/lr_scheduler.py [hunk 5]
```diff
@@ -779,15 +764,9 @@ class SequentialLR(LRScheduler):
 
     Example:
         >>> # xdoctest: +SKIP
-        >>> # Assuming optimizer uses lr = 1. for all groups
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:749]: same here

File: torch/optim/lr_scheduler.py [hunk 6]
```diff
@@ -922,13 +901,7 @@ class PolynomialLR(LRScheduler):
 
     Example:
         >>> # xdoctest: +SKIP("undefined vars")
-        >>> # Assuming optimizer uses lr = 0.001 for all groups
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:890]: same

File: torch/optim/lr_scheduler.py [hunk 7]
```diff
@@ -1083,13 +1056,7 @@ class ChainedScheduler(LRScheduler):
 
     Example:
         >>> # xdoctest: +SKIP
-        >>> # Assuming optimizer uses lr = 1. for all groups
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:1038]: same

**Comment by janeyx99** [comment]
The ChainedScheduler and Sequential should be the same, no?

**Review by janeyx99 — COMMENTED** [review]
thanks! looking better!

File: torch/optim/lr_scheduler.py (hunk 4 above)
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py:629]: Changed

File: torch/optim/lr_scheduler.py (hunk 5 above)
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py:749]: Changed

File: torch/optim/lr_scheduler.py (hunk 6 above)
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py:890]: Changed

File: torch/optim/lr_scheduler.py (hunk 7 above)
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py:1038]: Changed

File: torch/optim/lr_scheduler.py [hunk 8]
```diff
@@ -574,13 +571,7 @@ class ConstantLR(LRScheduler):
 
     Example:
         >>> # xdoctest: +SKIP
-        >>> # Assuming optimizer uses lr = 0.05 for all groups
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py]: Restored and changed numbers, thanks!

**Comment by zeshengzong** [comment]
> The ChainedScheduler and Sequential should be the same, no?

Not the same, `ChainedScheduler` execute each scheduler `step` method in a epoch, lr looks like a combination of multiple schedulers

https://github.com/pytorch/pytorch/blob/b01877aa1389a10bb3c17c9ecb8a79ff1c9a8f79/torch/optim/lr_scheduler.py#L1085-L1091

```python
>>> scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=20)
>>> scheduler2 = ExponentialLR(optimizer, gamma=0.9)
>>> scheduler = ChainedScheduler([scheduler1, scheduler2], optimizer=optimizer)
```

image


`Sequential` only execute one scheduler in an epoch, switch to another when get the milestone.

https://github.com/pytorch/pytorch/blob/b01877aa1389a10bb3c17c9ecb8a79ff1c9a8f79/torch/optim/lr_scheduler.py#L830-L840

```python
>>> scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=20)
>>> scheduler2 = ExponentialLR(optimizer, gamma=0.9)
>>> scheduler = SequentialLR(optimizer, schedulers=[scheduler1, scheduler2], milestones=[20])
```

image

**Review by janeyx99 — APPROVED** [review]
Thank you!
